File: Go/max_path_sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
    ret := math.MinInt32
    dfs(root, &ret)
    return ret
}

func dfs(root *TreeNode, ret *int) int {
    if root == nil {return 0}

    left := max(dfs(root.Left, ret), 0)
    right := max(dfs(root.Right, ret), 0)
    cur := root.Val + left + right
    *ret = max(*ret, cur)

    return root.Val + max(left, right)
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

File: Java/MaxPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        if(root == null) return 0;
        int[] max = new int[]{Integer.MIN_VALUE};

        dfs(root, max);
        return max[0];
    }

    private int dfs(TreeNode root, int[] max) {
        int left = root.left != null ? Math.max(dfs(root.left, max), 0): 0;
        int right = root.right != null ? Math.max(dfs(root.right, max), 0): 0;
        int cur = root.val + left + right;
        max[0] = Math.max(cur, max[0]);

        return root.val + Math.max(left, right);
    }
}